control-plane/csi-driver/src/bin/node/filesystem_ops.rs

```diff
@@ -0,0 +1,276 @@
+        static XFS: XFs = XFs {};
+        static UNSUPPORTED: Unsupported = Unsupported {};
+        match self.0 {
+            Fs::Ext4 => &EXT4FS,
+            Fs::Xfs => &XFS,
+            _ => &UNSUPPORTED,
+        }
+    }
+    /// Get a specific filesystem property by specifying the device path.
+    pub(crate) fn get_property(device: &str, property: &str) -> Result<String, DevInfoError> {
+        let probe = Probe::new_from_filename(device)?;
+        probe.do_probe()?;
+        probe.lookup_value(property)
+    }
+}
+
+#[async_trait]
+pub(crate) trait FileSystemOps: Send + Sync {
+    /// Create the filesystem using its fs util.
+    async fn create(&self, device: &str) -> Result<(), Error>;
+    /// Get the default mount options along with the user passed options for specific filesystems.
+    fn default_mount_flags(&self, mount_flags: Vec<String>) -> Vec<String>;
+    /// Unmount the filesystem if the filesystem uuid and the provided uuid differ.
+    fn unmount_on_fs_id_diff(
+        &self,
+        device_path: &str,
+        fs_staging_path: &str,
+        volume_uuid: &Uuid,
+    ) -> Result<(), String>;
+}
+
+#[async_trait]
+impl FileSystemOps for Ext4Fs {
+    async fn create(&self, device: &str) -> Result<(), Error> {
+        let binary = format!("mkfs.{}", "ext4");
+        let output = Command::new(&binary)
+            .arg(device)
+            .output()
+            .await
+            .map_err(|error| format!("failed to execute {binary}: {error}"))?;
+        ack_command_output(output, binary)
+    }
+
+    fn default_mount_flags(&self, mount_flags: Vec<String>) -> Vec<String> {
+        mount_flags
+    }
+
+    fn unmount_on_fs_id_diff(
+        &self,
+        _device_path: &str,
+        _fs_staging_path: &str,
+        _volume_uuid: &Uuid,
+    ) -> Result<(), String> {
+        Ok(())
+    }
+}
+
+impl Ext4Fs {
+    async fn repair(device: &str) -> Result<(), Error> {
+        let binary = "e2fsck".to_string();
+        let output = Command::new(&binary)
+            .arg("-y")
+            .arg("-f")
+            .arg(device)
+            .output()
+            .await
+            .map_err(|error| format!("failed to execute {binary}: {error}"))?;
+
+        trace!(
+            "Output from {} command: {}, status code: {:?}",
+            binary,
+            String::from_utf8(output.stdout.clone()).unwrap(),
+            output.status.code()
+        );
+
+        // For ext4 fs repair using e2fsck the success condition is when status code is 0 or 1.
+        if !output.status.success() && output.status.code() != Some(1) {
+            return Err(format!(
+                "{} command failed: {}",
+                binary,
+                String::from_utf8(output.stderr).unwrap()
+            ));
+        }
+        Ok(())
+    }
+
+    /// Change the uuid of the ext4 filesystem after repair.
+    pub(crate) async fn set_uuid(device: &str, uuid: &str) -> Result<(), Error> {
+        Self::repair(device).await?;
```

Comment:
First we should check if the current uuid is already the same, and if so don't bother doing anything.